client-config/src/main/java/com/palantir/conjure/java/client/config/ClientConfigurations.java

```diff
@@ -270,6 +298,59 @@ public String toString() {
         }
     }
 
+    private static final class NoProxyAwareSelector extends ProxySelector {
+        private static final ImmutableList<Proxy> NO_PROXIES = ImmutableList.of(Proxy.NO_PROXY);
+
+        private final ImmutableList<Proxy> proxies;
+        private final ImmutableSet<String> noProxyHosts; // host or domain suffixes
+
+        NoProxyAwareSelector(Proxy proxy, ImmutableSet<String> noProxyHosts) {
+            this.proxies = ImmutableList.of(proxy);
+            this.noProxyHosts = noProxyHosts;
+        }
+
+        @Override
+        public List<Proxy> select(URI uri) {
+            return shouldSkipProxy(uri) ? NO_PROXIES : proxies;
+        }
+
+        private boolean shouldSkipProxy(URI uri) {
+            String host = uri.getHost().toLowerCase(Locale.ROOT);
+            for (String noProxyHost : noProxyHosts) {
+                if (host.equals(noProxyHost) || host.endsWith("." + noProxyHost)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            if (this == other) {
+                return true;
+            }
+            if (other == null || NoProxyAwareSelector.class != other.getClass()) {
+                return false;
+            }
+            NoProxyAwareSelector that = (NoProxyAwareSelector) other;
+            return proxies.equals(that.proxies) && noProxyHosts.equals(that.noProxyHosts);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(proxies, noProxyHosts);
+        }
+
+        @Override
+        public String toString() {
+            return "NoProxyAwareProxySelector{proxy=" + proxies.stream().findFirst() + ", noProxyHosts=" + noProxyHosts
+                    + '}';
+        }
+
+        @Override
+        public void connectFailed(URI _uri, SocketAddress _sa, IOException _ioe) {}
```

Comment:
vnit/style: let's maybe move this method to just below the `select` or `shouldSkipProxy` methods, since it's also part of `ProxySelector` (also more consistent with `FixedProxySelector`)